### PR TITLE
Fix 404 when send pull request some situation 

### LINKED ISSUE #6302: compare/create PR produces a 404
- Gitea version (or commit ref): 1.7.3
- Git version: 2.21.0
- Operating system: CentOS 7
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [X] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [X] No
  - [ ] Not relevant
- Log gist:

## Description
when trying to create a PR gitea shows a 404.
Happens both on
- clicking on the "compare" button
- using the api https://myserver.example/api/v1/repos/myowner/myrepo/pulls

No relevant entries in the log files.

After downgrading to 1.7.2 everything works fine again.

### PR DESCRIPTION
**Pull request by lunny**
With integration tests and should fix #6302

### CODE REVIEW

File: routers/api/v1/repo/pull.go [hunk 1]
```diff
@@ -666,19 +666,37 @@ func parseCompareInfo(ctx *context.APIContext, form api.CreatePullRequestOption)
...
+	permBase, err := models.GetUserRepoPermission(baseRepo, ctx.User)
 	if err != nil {
 		ctx.ServerError("GetUserRepoPermission", err)
 		return nil, nil, nil, nil, "", ""
 	}
-	if !perm.CanReadIssuesOrPulls(true) {
+	if !permBase.CanReadIssuesOrPulls(true) || !permBase.CanRead(models.UnitTypeCode) {
 		if log.IsTrace() {
-			log.Trace("Permission Denied: User %-v cannot create/read pull requests in Repo %-v\nUser in headRepo has Permissions: %-+v",
+			log.Trace("Permission Denied: User %-v cannot create/read pull requests in Repo %-v\nUser in baseRepo has Permissions: %-+v",
```
Comment (zeripath) [inline routers/api/v1/repo/pull.go]:
```suggestion
			log.Trace("Permission Denied: User %-v cannot create/read pull requests or cannot read code in Repo %-v\nUser in baseRepo has Permissions: %-+v",
```

File: routers/api/v1/repo/pull.go [hunk 2]
```diff
@@ -666,19 +666,37 @@ func parseCompareInfo(ctx *context.APIContext, form api.CreatePullRequestOption)
...
+
+	// user should have permission to read headrepo's codes
+	permHead, err := models.GetUserRepoPermission(headRepo, ctx.User)
+	if err != nil {
+		ctx.ServerError("GetUserRepoPermission", err)
+		return nil, nil, nil, nil, "", ""
+	}
+	if !permHead.CanRead(models.UnitTypeCode) {
+		if log.IsTrace() {
+			log.Trace("Permission Denied: User: %-v cannot create/read pull requests in Repo: %-v\nUser in headRepo has Permissions: %-+v",
```
Comment (zeripath) [inline routers/api/v1/repo/pull.go]:
```suggestion
			log.Trace("Permission Denied: User: %-v cannot read code in Repo: %-v\nUser in headRepo has Permissions: %-+v",
```

File: routers/repo/pull.go [hunk 3]
```diff
@@ -710,17 +710,35 @@ func ParseCompareInfo(ctx *context.Context) (*models.User, *models.Repository, *
...
+	// user should have permission to read baseRepo's codes and pulls, NOT headRepo's
+	permBase, err := models.GetUserRepoPermission(baseRepo, ctx.User)
 	if err != nil {
 		ctx.ServerError("GetUserRepoPermission", err)
 		return nil, nil, nil, nil, "", ""
 	}
-	if !perm.CanReadIssuesOrPulls(true) {
+	if !permBase.CanReadIssuesOrPulls(true) || !permBase.CanRead(models.UnitTypeCode) {
+		if log.IsTrace() {
+			log.Trace("Permission Denied: User: %-v cannot create/read pull requests in Repo: %-v\nUser in baseRepo has Permissions: %-+v",
```
Comment (zeripath) [inline routers/repo/pull.go]:
```suggestion
			log.Trace("Permission Denied: User: %-v cannot create/read pull requests or cannot read code in Repo: %-v\nUser in baseRepo has Permissions: %-+v",
```

File: routers/repo/pull.go [hunk 4]
```diff
@@ -710,17 +710,35 @@ func ParseCompareInfo(ctx *context.Context) (*models.User, *models.Repository, *
...
+
+	// user should have permission to read headrepo's codes
+	permHead, err := models.GetUserRepoPermission(headRepo, ctx.User)
+	if err != nil {
+		ctx.ServerError("GetUserRepoPermission", err)
+		return nil, nil, nil, nil, "", ""
+	}
+	if !permHead.CanRead(models.UnitTypeCode) {
 		if log.IsTrace() {
 			log.Trace("Permission Denied: User: %-v cannot create/read pull requests in Repo: %-v\nUser in headRepo has Permissions: %-+v",
```
Comment (zeripath) [inline routers/repo/pull.go]:
```suggestion
			log.Trace("Permission Denied: User: %-v cannot read code requests in Repo: %-v\nUser in headRepo has Permissions: %-+v",
```

**Comment by zeripath** [comment]
@lunny, would it be possible to write a document somewhere explaining what the permissions should be doing things? (This isn't the PR for that I know but I think it's a good example of one where we're not very clear about what permissions things should have.)

**Review by zeripath — APPROVED** [review]

**Review by lafriks — APPROVED** [review]

**Comment by zeripath** [comment]
@lunny are you able to backport?

**Comment by lunny** [comment]
@zeripath I think that's a good idea. I will do that.

**Pull request merged**